tawazi/ops.py

```diff
@@ -53,40 +64,107 @@ def to_dag(
+
+                # 2.2 Construct non default arguments.
+                # Corresponding values must be provided during usage
+                args: List[ExecNode] = [ArgExecNode(_func, arg_name) for arg_name in func_args]
+                # 2.2 Construct Default arguments.
+                args.extend(
+                    [
+                        ArgExecNode(_func, arg_name, arg)
+                        for arg_name, arg in func_default_args.items()
+                    ]
+                )
+                # 2.3 Arguments are also ExecNodes that get executed inside the scheduler
+                node.exec_nodes.extend(args)
+
+                # 3. Execute the dependency describer function
+                # NOTE: Only ordered parameters are supported at the moment!
+                #  No **kwargs!! Only positional Arguments
+                returned_exec_nodes = _func(*args)
+
+                # 4. Construct the DAG instance
                 d = DAG(node.exec_nodes, max_concurrency=max_concurrency, behavior=behavior)
+
+            # clean up even in case an error is raised during dag construction
+            finally:
+                # 5. Clean global variable
+                # node.exec_nodes are deep copied inside the DAG.
+                #   we can empty the global variable node.exec_nodes
                 node.exec_nodes = []
 
-            return d
+            d.input_ids = [arg.id for arg in args]
+
+            # 6. make the return ids to be fetched at the end of the computation
+            # TODO: support Dicts and iterators etc.
+            err = TawaziTypeError(
+                "Return type of the pipeline must be either a Single value, Tuple of Values or Nothing"
+            )
+            # 6.1 returned values can be of multiple nature
+            return_ids: ReturnIDsType = []
+            # 6.2 No value returned by the execution
+            if returned_exec_nodes is None:
+                return_ids = None
+            # 6.3 a single value is returned
+            elif isinstance(returned_exec_nodes, ExecNode):
+                return_ids = returned_exec_nodes.id
+            # 6.4 multiple values returned
+            elif isinstance(returned_exec_nodes, (tuple, list)):
+                # 6.4.1 Collect all the return ids
+                for ren in returned_exec_nodes:
+                    if isinstance(ren, ExecNode):
+                        return_ids.append(ren.id)  # type: ignore
+                    else:
+                        # NOTE: this error shouldn't ever raise during usage.
+                        # Please report in https://github.com/mindee/tawazi/issues
+                        raise err
+                # 6.4.2 Cast to the corresponding type
+                if isinstance(returned_exec_nodes, tuple):
+                    return_ids = tuple(return_ids)  # type: ignore
+                # 6.4.3 No Cast is necessary for the List because this is the default
```

Comment:
same remark here, I think we should support the other native types in this release

Reply:
what are the other types that you think are crucial to support ?

Reply:
at least Dict and set, and maybe in the future TypeVar T to give full flexibilty but I doubt it's going to be easy, I'd suggest dict and set at the moment